sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.metrics;
+
+import com.google.common.base.Joiner;
+
+/**
+ * Standard IO Metrics.
+ */
+public class IOMetrics {
+
+  private static final String IO_NAMESPACE = "io";
+  private static final String IO_SPLITS_NAMESPACE = "io.splits";
+  private static final String SEPARATOR = ".";
+
+  private static final String ELEMENTS_READ = "elements_read";
+  private static final String BYTES_READ = "bytes_read";
+  private static final String BACKLOG_BYTES = "backlog_bytes";
+  private static final String BACKLOG_ELEMENTS = "backlog_elements";
+  private static final String ELEMENTS_WRITTEN = "elements_written";
+  private static final String BYTES_WRITTEN = "bytes_written";
+
+  private static final Counter ELEMENTS_READ_COUNTER = Metrics.counter(IO_NAMESPACE, ELEMENTS_READ);
+  private static final Counter BYTES_READ_COUNTER = Metrics.counter(IO_NAMESPACE, BYTES_READ);
+  private static final Counter ELEMENTS_WRITTEN_COUNTER =
+      Metrics.counter(IO_NAMESPACE, ELEMENTS_WRITTEN);
+  private static final Counter BYTES_WRITTEN_COUNTER =
+      Metrics.counter(IO_NAMESPACE, BYTES_WRITTEN);
+  private static final Gauge BACKLOG_BYTES_GAUGE = Metrics.gauge(IO_NAMESPACE, BACKLOG_BYTES);
+  private static final Gauge BACKLOG_ELEMENTS_GAUGE = Metrics.gauge(IO_NAMESPACE, BACKLOG_ELEMENTS);
+
+  // ------------------------- Source Metrics -------------------------
+
+  /**
+   * Counter of elements read by a source.
+   */
+  public static Counter elementsRead() {
```

Comment:
Splitting the factory seems reasonable -- then we also clearly distinguish the source and sink cases from each other, as well as from the case of reading/writing PCollections.